### PR TITLE
Wilson parallel upload

### LINKED ISSUE #2101: Allow for parallel export requests in wilson

### PR DESCRIPTION
**Pull request by domwst**
### Implemented parallel upload in wilson

...

### Changelog category 

* Not for changelog (changelog entry is not required)

### Additional information

...

### CODE REVIEW

File: ydb/library/actors/wilson/wilson_uploader.cpp [hunk 1]
```diff
@@ -115,11 +132,11 @@ namespace NWilson {
                 }
 
                 if (numSpansDropped) {
-                    LOG_ERROR_S(*TlsActivationContext, WILSON_SERVICE_ID,
+                    ALOG_ERROR(WILSON_SERVICE_ID,
                         "dropped " << numSpansDropped << " span(s) due to expiration");
                 }
 
-                if (Context || Spans.empty()) {
+                if (Spans.empty()) {
```
Comment (domwst) [inline ydb/library/actors/wilson/wilson_uploader.cpp:139]: Should we also limit upload parallelism?

File: ydb/library/actors/wilson/wilson_uploader.cpp [hunk 2]
```diff
@@ -267,29 +279,43 @@ namespace NWilson {
...
+                            "failed to commit traces: " << node->Status.error_message());
                     }
+                    
+                    --ExportRequestsCount;
+                    node->Unlink();
+                    delete node;
+                }
+
+                if (!ExportRequests.Empty()) {
+                    ScheduleWakeup(TDuration::MilliSeconds(100));
```
Comment (alexvru) [inline ydb/library/actors/wilson/wilson_uploader.cpp:318]: by the way, how will this work with a big rate of spans?
may be we need to run another thread polling CQ and sending messages to the uploader?

File: ydb/library/actors/wilson/wilson_uploader.cpp [hunk 3]
```diff
@@ -267,29 +279,43 @@ namespace NWilson {
...
-                    } else {
-                        ScheduleWakeup(TDuration::MilliSeconds(100));
+            void ReapCompletedRequests() {
+                if (ExportRequests.Empty()) {
+                    return;
+                }
+                void* tag;
+                bool ok;
+                while (CQ.AsyncNext(&tag, &ok, std::chrono::system_clock::now()) == grpc::CompletionQueue::GOT_EVENT) {
+                    auto node = static_cast<TExportRequestData*>(tag);
```
Comment (alexvru) [inline ydb/library/actors/wilson/wilson_uploader.cpp:305]: smart pointer

**Review by alexvru — CHANGES_REQUESTED** [review]

File: ydb/core/protos/config.proto [hunk 4]
```diff
@@ -1577,6 +1577,8 @@ message TTracingConfig {
         // time after which generated span will be discarded and will
         // not be sent to the collector
         optional uint32 SpanExportTimeoutSeconds = 5;
+        // maximum batch export requests being run simultaneously
```
[the remark this answers is not in the document]
Comment (domwst) [reply ydb/core/protos/config.proto:1580]: In requests. Each request exports one batch

[the remark this answers is not in the document]
Comment (domwst) [reply ydb/core/protos/config.proto:1580]: Renamed to MaxExportRequestsInflight

File: ydb/library/actors/wilson/wilson_uploader.cpp [hunk 5]
```diff
@@ -149,6 +157,10 @@ namespace NWilson {
                     ALOG_WARN(WILSON_SERVICE_ID, "max_spans_in_batch shold be greater than 0, changing to 1");
                     MaxSpansInBatch = 1;
                 }
+                if (MaxExportInflight == 0) {
+                    ALOG_WARN(WILSON_SERVICE_ID, "max_span_export_inflight should be greater than 0, channing to 1");
```
[the remark this answers is not in the document]
Comment (domwst) [reply ydb/library/actors/wilson/wilson_uploader.cpp]: Fixed

**Review by alexvru — APPROVED** [review]

**Pull request merged**